File: back-end/src/main/java/es/ull/project/domain/valueobject/capacity/VehicleCapacityKilograms.java
package es.ull.project.domain.valueobject.capacity;

import java.util.Objects;

/**
 * VehicleCapacityKilograms
 *
 * Represents the maximum capacity of a vehicle in kilograms.
 * Immutable value object that encapsulates the capacity value in kilograms.
 * The unit is always kilograms (fixed).
 */
public final class VehicleCapacityKilograms {

    /**
     * Error messages for validation.
     */
    private static final String ERROR_CAPACITY_NEGATIVE = "Vehicle capacity cannot be negative";
    private static final String ERROR_OTHER_CAPACITY_NULL = "Other VehicleCapacityKilograms cannot be null";
    private static final int ZERO = 0;

    /**
     * Required.
     * Capacity value in kilograms.
     * It is a required attribute and always measured in kilograms.
     */
    private final double kilograms;

    /**
     * Creates a new VehicleCapacityKilograms.
     *
     * @param kilograms Capacity value in kilograms (must be ≥ 0)
     * @throws IllegalArgumentException if kilograms is negative
     */
    public VehicleCapacityKilograms(double kilograms) {
        if (kilograms < ZERO) {
            throw new IllegalArgumentException(ERROR_CAPACITY_NEGATIVE);
        }
        this.kilograms = kilograms;
    }

    /**
     * Returns the capacity value in kilograms.
     *
     * @return Capacity value in kilograms
     */
    @com.fasterxml.jackson.annotation.JsonValue
    public double getKilograms() {
        return this.kilograms;
    }

    /**
     * Returns a new VehicleCapacityKilograms with the updated value.
     *
     * @param newKilograms New capacity value in kilograms
     * @return New VehicleCapacityKilograms
     * @throws IllegalArgumentException if newKilograms is negative
     */
    public VehicleCapacityKilograms setKilograms(double newKilograms) {
        return new VehicleCapacityKilograms(newKilograms);
    }

    /**
     * Compares if this capacity is greater than another.
     *
     * @param other Other VehicleCapacityKilograms
     * @return true if greater, false otherwise
     * @throws IllegalArgumentException if the other object is null
     */
    public boolean greaterThan(VehicleCapacityKilograms other) {
        if (other == null) {
            throw new IllegalArgumentException(ERROR_OTHER_CAPACITY_NULL);
        }
        return this.kilograms > other.kilograms;
    }

    /**
     * Compares if this capacity is greater than or equal to another.
     *
     * @param other Other VehicleCapacityKilograms
     * @return true if greater or equal, false otherwise
     * @throws IllegalArgumentException if the other object is null
     */
    public boolean greaterThanOrEqual(VehicleCapacityKilograms other) {
        if (other == null) {
            throw new IllegalArgumentException(ERROR_OTHER_CAPACITY_NULL);
        }
        return this.kilograms >= other.kilograms;
    }

    /**
     * Compares if this capacity is less than another.
     *
     * @param other Other VehicleCapacityKilograms
     * @return true if less, false otherwise
     * @throws IllegalArgumentException if the other object is null
     */
    public boolean lessThan(VehicleCapacityKilograms other) {
        if (other == null) {
            throw new IllegalArgumentException(ERROR_OTHER_CAPACITY_NULL);
        }
        return this.kilograms < other.kilograms;
    }

    /**
     * Compares equality of two VehicleCapacityKilograms instances.
     *
     * @param otherObject Other object to compare
     * @return true if equal, false otherwise
     */
    @Override
    public boolean equals(Object otherObject) {
        if (this == otherObject) {
            return true;
        }
        if (otherObject == null || getClass() != otherObject.getClass()) {
            return false;
        }
        VehicleCapacityKilograms other = (VehicleCapacityKilograms) otherObject;
        return Double.compare(this.kilograms, other.kilograms) == ZERO;
    }

    /**
     * Returns the hash code of the VehicleCapacityKilograms.
     *
     * @return hashCode
     */
    @Override
    public int hashCode() {
        return Objects.hash(kilograms);
    }

    /**
     * Returns the string representation of the VehicleCapacityKilograms.
     *
     * @return String
     */
    @Override
    public String toString() {
        return String.format("VehicleCapacityKilograms={kilograms=%s}", this.kilograms);
    }
}

File: front-end/src/domain/valueobject/capacity/vehicle-capacity-kilograms.ts
/**
 * VehicleCapacityKilograms
 *
 * Represents the maximum carrying capacity of a vehicle in kilograms.
 * Immutable value object that encapsulates the capacity value in kilograms.
 * The unit is always kilograms (fixed).
 */
export class VehicleCapacityKilograms {
  /**
   * Capacity value in kilograms.
   * Must be non-negative.
   */
  private readonly Kilograms: number;

  /**
   * Create a new `VehicleCapacityKilograms`.
   * @param Kilograms The capacity in kilograms (must be >= 0)
   * @throws Error if Kilograms is negative
   */
  constructor(Kilograms: number) {
    if (Kilograms < 0) {
      throw new Error('Vehicle capacity cannot be negative');
    }
    this.Kilograms = Kilograms;
  }

  /**
   * Get the capacity value in kilograms.
   * @returns The capacity in kilograms
   */
  getKilograms(): number {
    return this.Kilograms;
  }

  /**
   * Create a new instance with an updated value.
   * @param newKilograms The new capacity value in kilograms
   * @returns A new VehicleCapacityKilograms instance
   * @throws Error if newKilograms is negative
   */
  setKilograms(newKilograms: number): VehicleCapacityKilograms {
    return new VehicleCapacityKilograms(newKilograms);
  }

  /**
   * Compare if this capacity is greater than another.
   * @param other The other VehicleCapacityKilograms to compare
   * @returns true if this capacity is greater, false otherwise
   * @throws Error if other is null or undefined
   */
  greaterThan(other: VehicleCapacityKilograms): boolean {
    if (!other) {
      throw new Error('Other VehicleCapacityKilograms cannot be null');
    }
    return this.Kilograms > other.Kilograms;
  }

  /**
   * Compare if this capacity is greater than or equal to another.
   * @param other The other VehicleCapacityKilograms to compare
   * @returns true if this capacity is greater or equal, false otherwise
   * @throws Error if other is null or undefined
   */
  greaterThanOrEqual(other: VehicleCapacityKilograms): boolean {
    if (!other) {
      throw new Error('Other VehicleCapacityKilograms cannot be null');
    }
    return this.Kilograms >= other.Kilograms;
  }

  /**
   * Compare if this capacity is less than another.
   * @param other The other VehicleCapacityKilograms to compare
   * @returns true if this capacity is less, false otherwise
   * @throws Error if other is null or undefined
   */
  lessThan(other: VehicleCapacityKilograms): boolean {
    if (!other) {
      throw new Error('Other VehicleCapacityKilograms cannot be null');
    }
    return this.Kilograms < other.Kilograms;
  }

  /**
   * Compare equality with another object.
   * @param otherObject The object to compare with
   * @returns true if the objects are equal, false otherwise
   */
  equals(otherObject: unknown): boolean {
    if (this === otherObject) {
      return true;
    }
    if (!(otherObject instanceof VehicleCapacityKilograms)) {
      return false;
    }
    return this.Kilograms === otherObject.Kilograms;
  }

  /**
   * Get the hash code for this object.
   * @returns Hash code based on the kilograms value
   */
  hashCode(): number {
    return this.Kilograms;
  }

  /**
   * Get string representation.
   * @returns String representation of this capacity in kilograms
   */
  toString(): string {
    return `VehicleCapacityKilograms{Kilograms=${this.Kilograms}}`;
  }
}
